### PR TITLE
Fix conversions

### LINKED ISSUE #1296: Shifts should be allowed for base/shift operands of different bit-width and sign
Since commit 66011cb2c22a19c2fc18885d6a9f2a8fee191ce8 the following shader would fail to be converted to SPIR-V under OpenGL semantics (and I assume under Vulkan too):

> #extension GL_ARB_gpu_shader_int64 : require
> writeonly uniform image2D tex;
> layout(local_size_x = 54) in;
> uniform int64_t arg0;
> uniform uint arg1;
> uniform int64_t expected;
> 
> void main()
> {
>   vec4 tmp_color;
>   int64_t result = (arg0 << arg1);
>   tmp_color = result == expected ? vec4(0.0, 1.0, 0.0, 1.0) : vec4(1.0, 0.0, 0.0, 1.0);
>   ivec2 coord = ivec2(gl_GlobalInvocationID.xy);
>   imageStore(tex, coord, tmp_color);
> }

With the following error:

> '<<' :  wrong operand types: no operation '<<' exists that takes a left-hand operand of type 'layout( location=1) uniform int64_t' and a right operand of type 'layout( location=2) uniform uint' (or there is no acceptable conversion)

The mentioned commit changed the checks at TIntermediate::addConversion, forcing base/shift to be of the same sign and same rank. 

But from GLSL spec:

> The shift operators (<<) and (>>). For both operators, the operands must be signed or unsigned
> integers or integer vectors. One operand can be signed while the other is unsigned. In all cases, the
> resulting type will be the same type as the left operand. If the first operand is a scalar, the second
> operand has to be a scalar as well. If the first operand is a vector, the second operand must be a scalar
> or a vector, and the result is computed component-wise. The result is undefined if the right operand is
> negative, or greater than or equal to the number of bits in the left expression's base type.

so base/shift can be of different sign

And from SPIR-V spec:

> The type of each Base and Shift must be a scalar or vector of integer type. Base and Shift must have the same number
> of components. The number of components and bit width of the type of Base must be the same as in Result Type.
> Shift is treated as unsigned. The result is undefined if Shift is greater than the bit width of the components of Base.

it asks Base and Result to be the same bit-width, not Base and Shift operands.

This was found while working on improving ARB_gl_spirv support for the mesa driver. We use glslang to convert GLSL shaders to SPIR-V. After the mentioned commit, a test that was working properly using the generated SPIR-V started to fail as glslang rejected to generate such SPIR-V binary.

### PR DESCRIPTION
**Pull request by johnkslang**
This fixes #1296 and replaces #1297.

The above uncovered some subtle and complex interactions with the history of adding multiple-width types, while splitting addConversion(), in parallel with HLSL getting different treatment from GLSL.

This set of commits rationalizes all of this.

Of key importance is that types of binary operations are not simply symmetrical things to convert to each other the same way for all operations.  E.g., logical ops want bools, not matching conversions, and shifts want ints, not matches.

Also important is the split of addConversion() was incomplete, in that both copies still supported ops from the original, even though they didn't quite make sense in the copies, except that the unconverted HLSL path still exercised them (now inappropriately).

### CODE REVIEW

**Comment by johnkslang** [comment]
Any input here @alegal-arm?

File: glslang/MachineIndependent/Intermediate.cpp [hunk 1]
```diff
@@ -3087,8 +3007,10 @@ bool TIntermediate::promoteBinary(TIntermBinary& node)
         case EOpSub:
         case EOpDiv:
         case EOpMul:
-            left = addConversion(op, TType(EbtInt, EvqTemporary, left->getVectorSize()), left);
-            right = addConversion(op, TType(EbtInt, EvqTemporary, right->getVectorSize()), right);
+            if (left->getBasicType() == EbtBool)
```
Comment (alegal-arm) [inline glslang/MachineIndependent/Intermediate.cpp]: how can we get bool operand here?

Comment (johnkslang) [reply glslang/MachineIndependent/Intermediate.cpp]: This is an HLSL-specific path, which allows numeric operations on Boolean types.

Comment (johnkslang) [reply glslang/MachineIndependent/Intermediate.cpp]: Ah, this is the only reason addConversion was called, was to change int -> bool, and it was calling the split out unary/conversion addConversion, but with binary ops.  This helps clean all that up. (It is at result-promote time, not at operand convert time, even though it goes back and adjusts the operands.) It is done in two phases: first operand-operand matching, as the current binary addConversion does, and then promotion to rationalize all three types (operand, operand, and result type).

**Pull request merged**